File: backend/app/schemas/access.py
"""Pydantic schemas for access management endpoints."""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, EmailStr, field_validator

from app.core.email import normalize_email
from app.schemas.user import AccessScopeEnum, RoleRead


class PermissionRead(BaseModel):
    """Schema for permission details."""

    resource: str
    action: str
    description: Optional[str] = None

    model_config = {"from_attributes": True}


class RoleWithPermissions(BaseModel):
    """Role schema with its permissions."""

    id: int
    name: str
    display_name: str
    description: Optional[str] = None
    permissions: list[PermissionRead]

    model_config = {"from_attributes": True}


class AccessUserRead(BaseModel):
    """Schema for access management user list/read."""

    id: int
    email: str
    name: str
    is_active: bool
    role_id: int
    role: RoleRead
    department_id: Optional[int] = None
    department_name: Optional[str] = None
    manager_id: Optional[int] = None
    manager_name: Optional[str] = None
    access_scope: AccessScopeEnum
    scope_label: str
    effective_permissions: list[str]
    external_id: Optional[str] = None
    job_title: Optional[str] = None
    entra_business_role: Optional[str] = None
    directory_last_checked_at: datetime | None = None
    directory_last_seen_at: datetime | None = None
    directory_sync_status: Optional[str] = None
    deprovisioned_at: datetime | None = None
    deprovision_reason: Optional[str] = None
    capabilities: "AccessUserCapabilities | None" = None

    model_config = {"from_attributes": True}


class AccessUserCapabilities(BaseModel):
    """Backend-authoritative access-user action capabilities."""

    can_edit_identity: bool
    can_edit_business_access: bool
    can_edit_role: bool
    can_deactivate: bool
    can_change_active_status: bool
    can_break_glass_enable: bool
    can_revoke_sessions: bool


class AccessUserUpdate(BaseModel):
    """
    Schema for access-management updates.

    Backend policy:
    - name/email and Admin-role assignment are restricted to platform Admin
    - department/manager/scope and non-admin role assignment are restricted to CRO
    """

    name: Optional[str] = None
    email: Optional[EmailStr] = None
    role_id: Optional[int] = None
    department_id: Optional[int] = None
    manager_id: Optional[int] = None
    access_scope: Optional[AccessScopeEnum] = None

    @field_validator("email", mode="before")
    @classmethod
    def _normalize_email(cls, value: str | None) -> str | None:
        return normalize_email(value)

File: frontend/src/types/access.ts
/**
 * Access management types matching backend schemas.
 */

export type AccessScopeEnum = 'global' | 'department' | 'manager';

export interface PermissionRead {
    resource: string;
    action: string;
    description: string | null;
}

export interface RoleWithPermissions {
    id: number;
    name: string;
    display_name: string;
    description: string | null;
    permissions: PermissionRead[];
}

export interface AccessUserRead {
    id: number;
    email: string;
    name: string;
    is_active: boolean;
    local_suspended?: boolean;
    local_recovery_pending?: boolean;
    local_enrollment_state?: string | null;
    role_id: number;
    role: {
        id: number;
        name: string;
        display_name: string;
        description: string | null;
    };
    department_id: number | null;
    department_name: string | null;
    manager_id: number | null;
    manager_name: string | null;
    access_scope: AccessScopeEnum;
    scope_label: string;
    effective_permissions: string[];
    external_id?: string | null;
    job_title?: string | null;
    entra_business_role?: string | null;
    directory_last_checked_at?: string | null;
    directory_last_seen_at?: string | null;
    directory_sync_status?: string | null;
    deprovisioned_at?: string | null;
    deprovision_reason?: string | null;
    capabilities: AccessUserCapabilities;
}

export interface AccessUserCapabilities {
    can_edit_identity: boolean;
    can_edit_business_access: boolean;
    can_edit_role: boolean;
    can_deactivate: boolean;
    can_change_active_status: boolean;
    can_break_glass_enable: boolean;
    can_revoke_sessions: boolean;
    can_resume?: boolean;
    can_reissue_invitation?: boolean;
    can_cancel_invitation?: boolean;
    can_request_password_reset?: boolean;
    can_initiate_recovery?: boolean;
    recovery_offline_required?: boolean;
    can_check_directory?: boolean;
    active_status_block_reason?: string | null;
    directory_owned_fields?: string[];
    verified_identity_fields?: string[];
}

export interface AccessUserUpdate {
    name?: string;
    email?: string;
    role_id?: number;
    department_id?: number | null;
    manager_id?: number | null;
    access_scope?: AccessScopeEnum;
    is_active?: boolean;
}

export interface AccessUserFilters {
    department_id?: number;
    role_id?: number;
    access_scope?: AccessScopeEnum;
    is_privileged?: boolean;
}
